text/2022-05-12-multi-cluster-support.md

```diff
@@ -0,0 +1,328 @@
+notification… We should also provide a way to force remove both parent and child
+(maybe propagating the annotation is enough).
+
+The total workflow can be described in the following graph:
+
+```
+User creates chaos -> if this chaos contain a `remoteCluster` field, the remote chaos controller reconcile
+
+User delete chaos -> if this chaos contain a `remoteCluster` field, the remote chaos controller reconcile
+
+Chaos in child cluster changed -> the remote chaos controller reconcile (resources will be mapped to the chaos in parent cluster)
+
+Controller Reconcile -> If the Target Cluster is not ready, do nothing and return
+   -> If the chaos is being deleted
+      -> If the chaos in the child cluster doesn't exist, remove the finalizer
+      -> Apply and return
+   -> If the finalizer doesn't exist, add a finalizer
+   -> Get the chaos in child cluster
+      -> If the chaos in the child cluster doesn't exist, create the chaos in the child cluster
+      -> If the chaos in the child cluster exists, read the `Status` of the chaos in the child cluster
+         and update the `Status` of the chaos in the parent cluster
+```
+
+### Q&A
+
+1. Why should we install the chaos mesh controller in every cluster? Wouldn’t it
+   be better to manage all chaos in a single cluster (even process)?
+
+   It sounds great to listen and respond to every chaos in a single process,
+   which means we will only need to install chaos daemon in the target cluster.
+   However, the reality is that there is no common way to communicate with the
+   chaos daemon in another cluster. I have thought about proxy, port-forward or
+   other things, but nothing is reliable enough.
+
+   I also don’t want to depend on the user to tell us the network topology: the
+   clusters are in the same network / different network / proxy / ingress / … .
+   It’s too hard for both of us and users. If we operate the chaos through
+   another controller in the target cluster, we only need an available
+   kubeconfig, which is much simpler than the network configuration.
+
+2. Why is the `remoteCluster` an extra field? Would a standalone resource /
+   annotation / … be better?
+
+   I can’t say which one is better. It’s a blind choice. I don’t want to create
+   a new embedded standalone resource because the existing two embedded
+   resources: Schedule and Workflow are too huge (but it’s actually not a strong
+   evidence). I prefer fields over annotations because fields sound more like an
+   official API, but annotations do not (or maybe we could start from
+   annotations? I’m not sure).
+
+3. Can one chaos inject fault into different clusters?
+
+   Obviously, no. I don’t think we can provide a general way to implement this
+   in the near future.
+
+4. Where will the webhook run?
+
+   In the parent chaos mesh. Because we will not create chaos in the webhook
+   (but in the reconcile), the validation should be done in the parent’s
+   webhook.
```

Comment:
> I am not actually sure about that. I prefer to keep all the webhooks for both "ParentCluster" and "RemoteCluster" because of the possible version skew.

It's fine to run duplicated webhook multiple times (as our webhook doesn't have side-effect now :smiley: ), so it doesn't mean we couldn't run webhook in the remote cluster.

> And it seems we need to prevent users from creating *Chaos in RemoteCluster?

I don't think so. Creating *Chaos in RemoteCluster doesn't break the whole design. We have a label/annotation in the remote cluster to mark it as managed by the one in parent cluster (and it's already needed, for GC).